app/src/main/kotlin/com/bitwarden/authenticator/data/authenticator/manager/TotpCodeManagerImpl.kt

```diff
@@ -27,107 +22,62 @@ private const val ONE_SECOND_MILLISECOND = 1000L
-                scope = unconfinedScope,
-                started = SharingStarted.WhileSubscribed(),
-                initialValue = DataState.Loading,
-            )
+    override fun getTotpCodesFlow(
+        itemList: List<AuthenticatorItem>,
+    ): Flow<List<VerificationCodeItem>> {
+        val flows = itemList.map { it.toFlowOfVerificationCodes() }
+        return combine(flows) { it.toList() }
     }
 
-    override fun getTotpCodeStateFlow(
-        item: AuthenticatorItemEntity,
-    ): StateFlow<DataState<VerificationCodeItem?>> =
-        getTotpCodeStateFlowInternal(itemEntity = item)
-
-    @Suppress("LongMethod")
-    private fun getTotpCodeStateFlowInternal(
-        itemEntity: AuthenticatorItemEntity?,
-    ): StateFlow<DataState<VerificationCodeItem?>> {
-        val cipherId = itemEntity?.id ?: return MutableStateFlow(DataState.Loaded(null))
-
-        return mutableVerificationCodeStateFlowMap.getOrPut(itemEntity) {
-            flow<DataState<VerificationCodeItem?>> {
-                val totpCode = itemEntity.toOtpAuthUriString()
-
-                var item: VerificationCodeItem? = null
-                while (currentCoroutineContext().isActive) {
-                    val time = (clock.millis() / ONE_SECOND_MILLISECOND).toInt()
-                    val dateTime = DateTime.now()
-                    if (item == null || item.isExpired(clock = clock)) {
-                        authenticatorSdkSource
-                            .generateTotp(
-                                totp = totpCode,
-                                time = dateTime,
-                            )
-                            .onSuccess { response ->
-                                item = VerificationCodeItem(
-                                    code = response.code,
-                                    totpCode = totpCode,
-                                    periodSeconds = response.period.toInt(),
-                                    timeLeftSeconds = response.period.toInt() -
-                                        time % response.period.toInt(),
-                                    issueTime = clock.millis(),
-                                    id = cipherId,
-                                    username = itemEntity.accountName,
-                                    issuer = itemEntity.issuer,
-                                    favorite = itemEntity.favorite,
-                                )
-                            }
-                            .onFailure {
-                                emit(DataState.Loaded(null))
-                                return@flow
-                            }
-                    } else {
-                        item?.let {
-                            item = it.copy(
-                                timeLeftSeconds = it.periodSeconds -
-                                    (time % it.periodSeconds),
+    fun AuthenticatorItem.toFlowOfVerificationCodes(): Flow<VerificationCodeItem> {
```

Comment:
One thing that's nice here is that both the local codes and the shared codes get generated by this function. Two types of codes, one pipe!